import boto3
import time
import logging
from botocore.exceptions import NoCredentialsError, BotoCoreError
from core.config.settings import AWS_CONFIG

# Configurar logger
logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)
handler = logging.StreamHandler()
handler.setFormatter(logging.Formatter("%(asctime)s - %(levelname)s - %(message)s"))
logger.addHandler(handler)

class S3Client:
    def __init__(self, config=AWS_CONFIG):
        self.s3_client = None
        self.config = config
    

    def conn_s3(self):
        """
        Establece la conexi√≥n con S3 y devuelve la instancia del cliente.
        """
        try:
            self.s3_client = boto3.client(
                "s3",
                aws_access_key_id=self.config["ACCESS_KEY"],
                aws_secret_access_key=self.config["SECRET_KEY"]
            )
            print("‚úÖ Conexi√≥n a S3 establecida con √©xito.")
            return self  # üîπ Retorna el objeto `S3Client`, no solo `boto3.client`
        except Exception as e:
            print(f"‚ùå Error al conectar con S3: {e}")
            return None


    def list_buckets(self):
        """
        Lista los buckets de S3 disponibles si la conexi√≥n es exitosa.
        """
        if not self.s3_client:
            logger.error("‚ö†Ô∏è No hay una conexi√≥n activa a S3.")
            return None
        try:
            response = self.s3_client.list_buckets()
            buckets = [bucket["Name"] for bucket in response["Buckets"]]
            logger.info(f"üì¶ Buckets disponibles: {buckets}")
            return buckets
        except BotoCoreError as e:
            logger.error(f"‚ùå Error al listar los buckets de S3: {e}")
            return None
